File: server/src/main/java/dev/slimevr/protocol/pubsub/HashedTopicId.java
package dev.slimevr.protocol.pubsub;

import solarxr_protocol.pub_sub.TopicId;

import java.util.Objects;


// This class is so the HashMap referencing the TopicId as key works
// it needs a unique hashcode based on the topicId and also an equals function
// because equals hashcode does not mean equals strings
public class HashedTopicId {

	private final TopicId inner;
	private final int hashcode;

	public HashedTopicId(TopicId topicIdT) {
		this.inner = topicIdT;
		this.hashcode = (inner.getAppName()
			+ "."
			+ inner.getOrganization()
			+ "."
			+ inner.getTopic()).hashCode();
	}

	public TopicId getInner() {
		return inner;
	}

	@Override
	public int hashCode() {
		return hashcode;
	}

	@Override
	public boolean equals(Object o) {
		if (this == o)
			return true;
		if (o == null || getClass() != o.getClass())
			return false;
		HashedTopicId that = (HashedTopicId) o;
		return Objects.equals(inner.getOrganization(), that.getInner().getOrganization())
			&& Objects.equals(inner.getAppName(), that.getInner().getAppName())
			&& Objects.equals(inner.getTopic(), that.getInner().getTopic());
	}
}

File: server/core/src/main/java/dev/slimevr/protocol/pubsub/HashedTopicId.kt
package dev.slimevr.protocol.pubsub

import solarxr_protocol.pub_sub.TopicIdT

// This class is so the HashMap referencing the TopicId as key works
// it needs a unique hashcode based on the topicId and also an equals function
// because equals hashcode does not mean equals strings
class HashedTopicId(val inner: TopicIdT) {
	private val hashcode: Int = (
		(
			inner.appName +
				"." +
				inner.organization +
				"." +
				inner.topic
			)
		).hashCode()

	override fun hashCode(): Int = hashcode

	override fun equals(other: Any?): Boolean {
		if (this === other) return true
		if (other == null || javaClass != other.javaClass) return false
		val that = other as HashedTopicId
		return inner.organization == that.inner.organization &&
			inner.appName == that.inner.appName &&
			inner.topic == that.inner.topic
	}
}
